File: src/DIMEX.c
#include <stdlib.h>
#include <string.h>
#include <stdio.h>
#include <semaphore.h>
#include "../include/chan.h"
#include "../include/p2plink.h"
#include "../include/DIMEX.h"

DIMEX_Module*
__new_dimex(char** addresses, unsigned char addresses_len, int id){
    char* address = addresses[id];
    //extract port my number
    //TODO: probably it's going to be necessary to extract the port from 
    // the address in the form 127.0.0.1:9000
    PP2PLink* p2p = new_p2p_link(1,"9000");
    chan*     Req = start_chan(1, (void**)calloc(1, sizeof(dmxReq*)));
    chan*     Ind = start_chan(1, (void**)calloc(1, sizeof(dmxResp*)));
    char** waiting = (char**)calloc(addresses_len,sizeof(char*));
    sem_t* global_mutex = (sem_t*)calloc(1,sizeof(sem_t));
    sem_init(global_mutex,0,1);
    for (size_t i = 0; i < addresses_len; i++){
        waiting[i] = calloc(1,sizeof(char));
        *waiting[i] = 0;
    };
    DIMEX_Module* dimex = (DIMEX_Module*)calloc(1,sizeof(DIMEX_Module));
    dimex->Req           = Req;
    dimex->Ind           = Ind;
    dimex->addresses     = addresses;
    dimex->addresses_len = addresses_len;
    dimex->id            = id;
    dimex->st            = noMX;
    dimex->lcl           = 0;
    dimex->req_ts        = 0;
    dimex->nbr_resps     = 0;
    dimex->waiting       = waiting;
    dimex->p2p           = p2p;
    dimex->global_mutex  = global_mutex;
    return dimex;
};
DIMEX_Module*
new_dimex(char** addresses, unsigned char addresses_len, int id){
    DIMEX_Module* dimex =__new_dimex(addresses, addresses_len, id);
    start_dimex(dimex);
    return dimex;
}
void*
handle_msg_from_app(DIMEX_Module* module){
    while(1){
        dmxReq* dmxR = receive_data(module->Req);
        sem_wait(module->global_mutex);
        switch (*dmxR){
            case ENTER:
                printf("app request mutex - process id%d\n",module->id);
                //handleUponReqEntry()
                break;
            case EXIT:
                //handleUponReqExit() 
                break;
        
        default:
            break;
        };
        sem_post(module->global_mutex);
        if(dmxR){
            free(dmxR);
        };
    }

}
void*
handle_peer_msg(DIMEX_Module* module){
    PP2PLink_Ind_Message* msg_outro = NULL;
while (1){
    msg_outro = receive_data(module->p2p->ind);
    if(!msg_outro){
        msg_outro = NULL;
        continue;
    };
    HANDLE_NULL_ALLOC(msg_outro->from,cleanup_msg_ptr);
    HANDLE_NULL_ALLOC(msg_outro->message,cleanup_msg_ptr);
    sem_wait(module->global_mutex);
    if(strstr(msg_outro->message,"respOK")){
        //handleUponDeliverRespOk
    }else if(strstr(msg_outro->message,"reqEntry")){
        //handleUponDeliverReqEntry
    }else{
        CLEANUP(free_mem);
    }
    sem_post(module->global_mutex);
    };
    cleanup_msg_ptr:
        if(msg_outro->message) free(msg_outro->message);
        if(msg_outro->from)    free(msg_outro->from);
        if(msg_outro)          free(msg_outro);
    free_mem:
        if(msg_outro->message) free(msg_outro->message);
        if(msg_outro->from)    free(msg_outro->from);
        if(msg_outro)          free(msg_outro);
}
void
start_dimex(DIMEX_Module* module){
    // Create two threads, one to handle messages from the app
    pthread_t* handle_app_th  = NULL;
    pthread_t* handle_peer_msg_th = NULL;
    handle_app_th = (pthread_t*)calloc(1,sizeof(pthread_t));
    handle_peer_msg_th = (pthread_t*)calloc(1,sizeof(pthread_t));
    HANDLE_NULL_ALLOC(handle_app_th, cleanup_start_dimex);
    HANDLE_NULL_ALLOC(handle_peer_msg_th, cleanup_start_dimex);
    if(pthread_create(handle_app_th, NULL, handle_msg_from_app, module)!=0){
        CLEANUP(cleanup_start_dimex);
    };
    // another to handle messages coming from other processes
    if(pthread_create(handle_peer_msg_th, NULL, handle_peer_msg, module)!=0){
        CLEANUP(cleanup_start_dimex);
    };
    cleanup_start_dimex:
        if(handle_app_th) free(handle_app_th);
        if(handle_peer_msg_th) free(handle_peer_msg_th);
        fprintf(stderr, "Error: failed to start dimex\n");
        exit(1);
};

void
send_to_link(DIMEX_Module* dimex, char* address, char* content){
    PP2PLink_Req_Message* msg = init_p2p_req(address, content);
    send_data(dimex->p2p->req, msg);
    printf("Warning: message sent from process %d to %s\n", dimex->id, content);
};

void
handle_upon_req_entry(DIMEX_Module* module){
    module->lcl++;
    module->req_ts    = module->lcl;
    module->nbr_resps = 0;
    char* message     = NULL;
    char* entry_msg   = "reqEntry";
    unsigned int entry_msg_len = strlen(entry_msg);
    char* pipe        = "|";
    int   pipe_len    = 1;
    char  id_str[6]   = {0};
    sprintf(id_str, "%d", module->id);
    unsigned int id_str_len = strlen(id_str);
    char req_ts_str[20] = {0};
    sprintf(req_ts_str, "%d", module->req_ts);
    char req_ts_str_len = strlen(req_ts_str);
    int msg_len = entry_msg_len + pipe_len + id_str_len + pipe_len + req_ts_str_len + 1;
    for (size_t i = 0; i < module->addresses_len; i++){
        if(i!=module->id){
            char* message = (char*) calloc(msg_len,sizeof(char));
            sprintf(message, "%s|%s|%s", entry_msg, id_str, req_ts_str);
            send_to_link(module,module->addresses[i],message);
        };
    };
    module->st = wantMX;
    return;
}

File: src/DIMEX.go
package DIMEX

import (
	PP2PLink "SD/PP2PLink"
	"fmt"
	"strconv"
	"strings"
)

// ------------------------------------------------------------------------------------
// ------- principais tipos
// ------------------------------------------------------------------------------------

type State int // enumeracao dos estados possiveis de um processo
const (
	noMX State = iota
	wantMX
	inMX
)

type dmxReq int // enumeracao dos estados possiveis de um processo
const (
	ENTER dmxReq = iota
	EXIT
)

type dmxResp struct { // mensagem do módulo DIMEX infrmando que pode acessar - pode ser somente um sinal (vazio)
	// mensagem para aplicacao indicando que pode prosseguir
}

type DIMEX_Module struct {
	Req       chan dmxReq  // canal para receber pedidos da aplicacao (REQ e EXIT)
	Ind       chan dmxResp // canal para informar aplicacao que pode acessar
	addresses []string     // endereco de todos, na mesma ordem
	id        int          // identificador do processo - é o indice no array de enderecos acima
	st        State        // estado deste processo na exclusao mutua distribuida
	waiting   []bool       // processos aguardando tem flag true
	lcl       int          // relogio logico local
	reqTs     int          // timestamp local da ultima requisicao deste processo
	nbrResps  int
	dbg       bool

	Pp2plink *PP2PLink.PP2PLink // acesso aa comunicacao enviar por PP2PLinq.Req  e receber por PP2PLinq.Ind
}

// ------------------------------------------------------------------------------------
// ------- inicializacao
// ------------------------------------------------------------------------------------

func NewDIMEX(_addresses []string, _id int, _dbg bool) *DIMEX_Module {

	p2p := PP2PLink.NewPP2PLink(_addresses[_id], _dbg)

	dmx := &DIMEX_Module{
		Req: make(chan dmxReq, 1),
		Ind: make(chan dmxResp, 1),

		addresses: _addresses,
		id:        _id,
		st:        noMX,
		waiting:   make([]bool, len(_addresses)),
		lcl:       0,
		reqTs:     0,
		dbg:       _dbg,

		Pp2plink: p2p}

	for i := 0; i < len(dmx.waiting); i++ {
		dmx.waiting[i] = false
	}
	dmx.Start()
	dmx.outDbg("Init DIMEX!")
	return dmx
}

// ------------------------------------------------------------------------------------
// ------- nucleo do funcionamento
// ------------------------------------------------------------------------------------

func (module *DIMEX_Module) Start() {

	go func() {
		for {
			select {
			case dmxR := <-module.Req: // vindo da  aplicação
				if dmxR == ENTER {
					module.outDbg("app pede mx")
					module.handleUponReqEntry() // ENTRADA DO ALGORITMO

				} else if dmxR == EXIT {
					module.outDbg("app libera mx")
					module.handleUponReqExit() // ENTRADA DO ALGORITMO
				}

			case msgOutro := <-module.Pp2plink.Ind: // vindo de outro processo
				//fmt.Printf("dimex recebe da rede: ", msgOutro)
				if strings.Contains(msgOutro.Message, "respOK") {
					module.outDbg("         <<<---- responde! " + msgOutro.Message)
					module.handleUponDeliverRespOk(msgOutro) // ENTRADA DO ALGORITMO

				} else if strings.Contains(msgOutro.Message, "reqEntry") {
					module.outDbg("          <<<---- pede??  " + msgOutro.Message)
					module.handleUponDeliverReqEntry(msgOutro) // ENTRADA DO ALGORITMO

				}
			}
		}
	}()
}

// ------------------------------------------------------------------------------------
// ------- tratamento de pedidos vindos da aplicacao
// ------- UPON ENTRY
// ------- UPON EXIT
// ------------------------------------------------------------------------------------

// handleUponReqEntry é uma função do módulo DIMEX que lida com as solicitações de
// entrada em uma seção crítica.
// A função incrementa o timestamp lógico do módulo, armazena o timestamp da solicitação
// e zera a contagem de respostas.
// Em seguida, a função envia solicitações para todos os outros processos e
// atualiza o estado do módulo para 'querendo entrar na seção crítica'.
func (module *DIMEX_Module) handleUponReqEntry() {
	message := ""
	space := "|"
	module.lcl++
	module.reqTs = module.lcl
	module.nbrResps = 0
	for i := 0; i < len(module.addresses); i++ {
		if i != module.id {
			message = "reqEntry" + space + strconv.Itoa(module.id) + space + strconv.Itoa(module.reqTs)
			module.sendToLink(module.addresses[i], message, space)
		}
	}
	module.st = wantMX
}

// handleUponReqExit é uma função do módulo DIMEX que lida com as solicitações de saída de uma seção crítica.
// A função envia uma resposta de sucesso para todos os processos que estavam esperando para entrar na seção crítica.
// Em seguida, a função atualiza o estado do módulo para 'não na seção crítica' e zera a lista de processos esperando.
func (module *DIMEX_Module) handleUponReqExit() {
	space := "|" // separador de campos
	for i := 0; i < len(module.waiting); i++ {
		if module.waiting[i] {
			module.sendToLink(module.addresses[i], "respOK"+space+strconv.Itoa(module.id), space)
			module.waiting[i] = false
		}
	}
	module.st = noMX
}

// ------------------------------------------------------------------------------------
// ------- tratamento de mensagens de outros processos
// ------- UPON respOK
// ------- UPON reqEntry
// ------------------------------------------------------------------------------------

// handleUponDeliverRespOk é uma função do módulo DIMEX que lida com as respostas
// de sucesso recebidas de outros processos.
// A função incrementa o contador de respostas e, se todas as respostas forem recebidas,
// envia uma resposta para o canal Ind do módulo e atualiza o estado do módulo para 'dentro da seção crítica'.
func (module *DIMEX_Module) handleUponDeliverRespOk(msgOutro PP2PLink.PP2PLink_Ind_Message) {
	module.nbrResps++
	if module.nbrResps == len(module.addresses)-1 {
		module.Ind <- dmxResp{}
		module.st = inMX
	}
}

// handleUponDeliverReqEntry é uma função do módulo DIMEX que lida com as solicitações
// de entrada recebidas de outros processos. A função analisa a mensagem recebida,
// verifica o estado do módulo e o timestamp da solicitação, e envia uma resposta apropriada.
// Se o módulo estiver 'querendo entrar na seção crítica' e o timestamp da solicitação
// for menor que o do módulo, ou se o módulo estiver 'não na seção crítica',
// a função envia uma resposta de sucesso. Caso contrário, o ID do processo é
// adicionado à lista de processos esperando. Se o timestamp da solicitação for
// maior que o timestamp lógico do módulo, o timestamp lógico do módulo é atualizado.
func (module *DIMEX_Module) handleUponDeliverReqEntry(msgOutro PP2PLink.PP2PLink_Ind_Message) {
	input := msgOutro.Message
	parts := strings.Split(input, "|")
	otID, _ := strconv.Atoi(parts[1])
	otRts, _ := strconv.Atoi(parts[2])
	if (module.st == wantMX && !before(module.id, module.reqTs, otID, otRts)) || module.st == noMX {
		module.sendToLink(module.addresses[otID], "respOK", "|")
	} else {
		module.waiting[otID] = true
	}
	if otRts > module.lcl {
		module.lcl = otRts
	}
}

// ------------------------------------------------------------------------------------
// ------- funcoes de ajuda
// ------------------------------------------------------------------------------------

func (module *DIMEX_Module) sendToLink(address string, content string, space string) {
	module.outDbg(space + " ---->>>>   to: " + address + "     msg: " + content)
	module.Pp2plink.Req <- PP2PLink.PP2PLink_Req_Message{
		To:      address,
		Message: content}
}

func before(oneId, oneTs, othId, othTs int) bool {
	if oneTs < othTs {
		return true
	} else if oneTs > othTs {
		return false
	} else {
		return oneId < othId
	}
}

func (module *DIMEX_Module) outDbg(s string) {
	if module.dbg {
		fmt.Println(". . . . . . . . . . . . [ DIMEX : " + s + " ]")
	}
}
